src/librustc_typeck/collect.rs

```diff
@@ -1377,42 +1377,46 @@ pub fn checked_type_of<'a, 'tcx>(
                                 }
                                 bug!("no arg matching AnonConst in path")
                             }
-                            match path.res {
-                                // We've encountered an `AnonConst` in some path, so we need to
-                                // figure out which generic parameter it corresponds to and return
-                                // the relevant type.
+
+                            // We've encountered an `AnonConst` in some path, so we need to
+                            // figure out which generic parameter it corresponds to and return
+                            // the relevant type.
+                            let generics = match path.res {
+                                Res::Def(DefKind::Ctor(CtorOf::Struct, _), def_id) =>
+                                    tcx.generics_of(tcx.parent(def_id).unwrap()),
                                 Res::Def(DefKind::Struct, def_id)
                                 | Res::Def(DefKind::Union, def_id)
                                 | Res::Def(DefKind::Enum, def_id)
-                                | Res::Def(DefKind::Fn, def_id) => {
-                                    let generics = tcx.generics_of(def_id);
-                                    let mut param_index = 0;
-                                    for param in &generics.params {
-                                        if let ty::GenericParamDefKind::Const = param.kind {
-                                            if param_index == arg_index {
-                                                return Some(tcx.type_of(param.def_id));
-                                            }
-                                            param_index += 1;
-                                        }
-                                    }
-                                    // This is no generic parameter associated with the arg. This is
-                                    // probably from an extra arg where one is not needed.
-                                    return Some(tcx.types.err);
-                                }
-                                Res::Err => tcx.types.err,
+                                | Res::Def(DefKind::Fn, def_id) =>
```

Comment:
I think we should just collapse all the above arms into `Res::Def(_, def_id) => tcx.generics_of(def_id)`: it means we won't accidentally miss any more cases in the future.